### PR TITLE
add haps configuration (cycle lr schedule)

### PR DESCRIPTION
**Pull request by blahBlahhhJ**
Add inverse decay schedule and haps configuration.

set config `haps` to a list of step points at which learning rate should decay to minimum.

The schedule structure now becomes
```
[warmup] - {[stable] - [decay]} x len(haps) - [cooldown]
```

For example, the following plot shows the schedule of this configuration
```
self.learning_rate = 6e-4
self.min_lr_ratio = 0.1

self.warmup_ratio = None
self.warmup = 0.1
self.stable = 0.8
self.cooldown = 0.1
self.lr_schedule = "inv"
self.haps = [3e5, 6e5]
```

Note that `self.stable = 0.8` means within every haps cycle (e.g. 1e5~3e5, 3e5~6e5, 6e5~9e5), 80% of the steps is in stable phase and 20% of the steps is in decay phase.

image

### CODE REVIEW

File: src/levanter/optim/config.py [hunk 1]
```diff
@@ -24,14 +24,16 @@ class OptimizerConfig(draccus.ChoiceRegistry, abc.ABC):
...
+    """The lr scheduler operates on 4 stages: [warmup] - {[stable] - [decay]} x haps - [cooldown]"""
     warmup: float = 0.01
     """fraction of training steps to use as warmup, or steps to use. 0.0 means no warmup"""
     stable: float = 0.00
     """fraction of training steps to use as stable, or steps to use. 0.0 means no stable"""
     cooldown: float = 0.0
     """fraction of training steps to use as cooldown, or steps to use. 0.0 means no cooldown"""
     lr_schedule: str = "cosine"  # constant, cosine, linear
+    haps: Optional[list[int]] = None
+    """list of integers indicating pit stop steps."""
```
Comment (dlwh) [inline src/levanter/optim/config.py:36]: add a link to the arxiv paper?

File: src/levanter/optim/config.py [hunk 2]
```diff
@@ -215,9 +235,7 @@ class HessianOptConfig(OptimizerConfig, abc.ABC):
 class AdamConfig(OptimizerConfig):
     weight_decay: float = 0.1
     beta1: float = 0.9
-    # cf https://docs.mosaicml.com/projects/composer/en/latest/api_reference/generated/composer.optim.DecoupledAdamW.html
-    # https://x.com/giffmana/status/1692641748445438301
-    beta2: float = 0.95
+    beta2: float = 0.999
     epsilon: float = 1e-8
```
Comment (dlwh) [inline src/levanter/optim/config.py:239]: revert

**Comment by dlwh** [comment]
should we maybe call the config pit_stops instead?

**Comment by dlwh** [comment]
(test failure was just insufficient TPU resources)

**Review by dlwh — APPROVED** [review]
